### PR TITLE
Save argument properties in argument object

### LINKED ISSUE #1331: Move inout tag to FunctionDefArgument
The class `FunctionDefArgument` was introduced in August last year. It contains many useful information about the argument of a function definition which are extracted from the original python code, however it does not contain all of the information which is useful for lower level languages. Instead this information is stored nearby. This is notably the case for the `FunctionDef.arguments_inout` property which provides additional information about whether the `FunctionDefArgument` is modified by the function. This property should be deleted and the constness should be made evident from `FunctionDefArgument`

### PR DESCRIPTION
**Pull request by EmilyBourne**
Use the `FunctionDefArgument` object to save information about whether an argument is modified by a function or not. This replaces the `FunctionDef.arguments_inout` property and simplifies the code slightly. Fixes #1331

### CODE REVIEW

File: pyccel/parser/semantic.py [hunk 1]
```diff
@@ -3275,50 +3290,23 @@ def _visit_FunctionDef(self, expr, **settings):
 
             # Collect the modified objects
             lhs_assigns   = [a.lhs for a in assigns]
-            modified_args = [func_arg for f in calls
-                                for func_arg, inout in zip(f.args,f.funcdef.arguments_inout) if inout]
+            modified_args = [call_arg.value for f in calls
+                                for call_arg, func_arg in zip(f.args,f.funcdef.arguments) if func_arg.inout]
```
Comment (yguclu) [inline pyccel/parser/semantic.py:3294]: Can you please add a comma between the arguments of `zip`?

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:3294]: Do you mean a space?

Comment (yguclu) [reply pyccel/parser/semantic.py:3294]: Yes! I meant a space after the comma

Comment (EmilyBourne) [reply pyccel/parser/semantic.py]: Fixed b80b62c3f6a4032a8edfc87cbc1416456e96faa5

File: pyccel/parser/semantic.py [hunk 2]
```diff
@@ -170,16 +170,36 @@ def _get_name(var):
...
+        A list of parsers describing the files which import this file.
+
+    d_parsers : list
+        A list of parsers describing files imported by this file.
+
+    **kwargs : dict
+        Additional keyword arguments for BasicParser.
+    """
+
+    def __init__(self, inputs, parents = (), d_parsers = (), **kwargs):
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/parser/semantic.py]: Fixed b80b62c3f6a4032a8edfc87cbc1416456e96faa5

**Review by yguclu — APPROVED** [review]

**Pull request merged**